axiom/nr_php_packages.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "nr_axiom.h"
+#include "nr_limits.h"
+#include "nr_php_packages.h"
+#include "util_memory.h"
+#include "util_minmax_heap.h"
+#include "util_vector.h"
+#include "util_hashmap.h"
+#include "util_hashmap_private.h"
+#include "util_strings.h"
+
+nr_php_package_t* nr_php_package_create(char* name, char* version) {
+  nr_php_package_t* p = NULL;
+
+  if (NULL == name || NULL == version) {
+    return NULL;
+  }
+
+  p = (nr_php_package_t*)nr_malloc(sizeof(nr_php_package_t));
+  if (NULL == p) {
+    return NULL;
+  }
+
+  p->package_name = nr_strdup(name);
+  p->package_version = nr_strdup(version);
+
+  return p;
+}
+
+void nr_php_package_destroy(nr_php_package_t* p) {
+  if (NULL != p) {
+    nr_free(p->package_name);
+    nr_free(p->package_version);
+    nr_free(p);
+  }
+}
+
+void nr_php_packages_add_package(nr_php_packages_t** h, nr_php_package_t* p) {
+  if (NULL == h) {
+    return;
+  }
+
+  if (NULL == p || NULL == p->package_name || NULL == p->package_version) {
+    return;
+  }
+
+  if (NULL == *h) {
+    *h = nr_hashmap_create((nr_hashmap_dtor_func_t)nr_php_package_destroy);
+  }
+
+  // If package with the same key already exists, we will check if the value is
+  // different. If it is different, then we will update the value of the package
+  if (0
+      != nr_php_packages_has_package(*h, p->package_name,
+                                     nr_strlen(p->package_name))) {
+    nr_php_package_t* package = (nr_php_package_t*)nr_hashmap_get(
+        *h, p->package_name, nr_strlen(p->package_name));
+    if (NULL != package) {
+      if (0 != nr_strcmp(package->package_version, p->package_version)) {
+        nr_free(package->package_version);
+        package->package_version = nr_strdup(p->package_version);
+      }
+    }
+    nr_php_package_destroy(p);
+    return;
+  }
+
+  nr_hashmap_set(*h, p->package_name, nr_strlen(p->package_name), p);
+}
+
+void nr_php_packages_destroy(nr_php_packages_t** h) {
```

Comment:
Pointers to pointers should be avoided to prevent unnecessary abstraction that can lead to confusion. Rather than using a pointer to a pointer, just use the pointer value. When you need to call `nr_hashmap_destroy`, just pass it `&h` instead.

Reply:
I think double pointer is necessary to ensure there are no dangling pointers. However, since this function is just wrapping an existing function, I will remove it and directly call nr_hashmap_destroy instead

Reply:
I think leaving the wrapper function makes sense as it completes the nr_php_packages API.